lib/omniauth/strategies/paypal_permissions.rb

```diff
@@ -110,7 +110,15 @@ def load_identity
           identity["display_name"] = identity["company_name"] || identity["full_name"]
           identity
         else
-          fail!(:invalid_credentials)
+          error = response.error.first
+          if error.present?
+            {
```

Comment:
Instead of returning a hash, I think it'd be better to define an exception class (say PaypalPermissions::Error), and then call fail!(:invalid_credentials, PaypalPermissions::Error.new(error.try(:errorId), error.try(:message)))

That way the failure will pass through omniauth's failure callback flow, but instead of raising the generic Omniauth::Error, you can check for PaypalPermissions::Error.

Does that make sense?